### PR TITLE
Resolve buffer underrun issue by adding emitQueue

### PR DESCRIPTION
**Pull request by ajpierce**
Github user @heberallred gets the credit for solving this one!  Addresses issue #8.

From his fork I checked out `lib/incoming.js` and `lib/socket.js`, cleaned up the code a little bit, and then tested it against the reqHistoricalData test script (posted as a Gist in the comments of #8) that always resulted in buffer underrun errors.

The result is that I'm now consistently getting all historical data with no "Unknown first token" errors or lost packets! If somebody else has a more extensive test suite, I'd really appreciate another pair of eyes on this one.  So far, it looks like this issue has been resolved!

Thanks, @heberallred!

### CODE REVIEW

**Comment by pilwon** [comment]
Thanks @heberallred for your fix and @ajpierce for putting it together to submit this PR.

@ajpierce Just a couple of suggestions before your awesome PR gets merged:
* Turn `UnderrunError` into a custom error type extending the standard Error type. Here's the [reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#Custom_Error_Types)) and feel free to add another file (ex: `lib/errors.js`) if you find it better.
* Remove `console.log(...)` lines used for debugging purposes.

**Comment by ajpierce** [comment]
@pilwon `console.log` statements are gone! I'll have the `UnderrunError` fix pushed up by the end of the day :+1:

**Comment by pilwon** [comment]
@ajpierce Ah my bad. I only looked at your first commit :+1:

**Comment by ajpierce** [comment]
Code is ready for re-review!

I kept the file headers consistent with what I saw in the rest of the application, but I changed the export statement a little to allow for multiple custom error messages in the future.

Let me know if there's anything else that needs changing!

**Comment by pilwon** [comment]
@ajpierce Thanks- v0.1.7 contains this PR.

File: lib/socket.js [hunk 1]
```diff
@@ -21,19 +21,26 @@ function Socket(controller) {
...
+  this.dataFragment = '';  // Prevents partial packets from getting lost (added by heberallred)
 }
 
 Socket.prototype._onConnect = function () {
   this._connected = true;
   this._controller.emit('connected');
 
-  this._controller.run('sendAsync', [C.CLIENT_VERSION]);
-  this._controller.run('sendAsync', [this._controller.options.clientId]);
+  this._controller.run('sendAsync', [C.CLIENT_VERSION, this._controller.options.clientId]);
```
Comment (ajpierce) [inline lib/socket.js]: @lewisdawson I'm not sure why these were combined, but let's break them back out into two calls again and see what happens

**Comment by nemozny** [comment]
Hi, I just had
`uncaughtException: An underrun error has occurred`

What I did:
1. I caught error 1100 - "Connectivity between IB and Trader Workstation has been lost."
2. The code following the error:
```
ib.disconnect() // .on("connected" was true    
     .connect()  // .on("disconnected" was true    
     .reqIds(1);
```     
Nothing else.
But I don't have a log line on reqIds(1), so I don't know at which point exactly the error happened.

What is the correct handling of "lost connectivity"? For now I just added a timeout between disconnect and connect.

**Pull request merged**